{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA input:  ACGATACGCCGACATCGGACGTTT\n",
      "Complement:  TGCTATGCGGCTGTAGCCTGCAAA\n",
      "mRNA:  UGCUAUGCGGCUGUAGCCUGCAAA\n",
      "Aminoacid:  Cys (C) - Tyr (Y) - Ala (A) - Ala (A) - Val (V) - Ala (A) - Cys (C) - Lys (K)\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Translate a DNA sequence ->  mRNA (using 'U' instead of 'Thr (T) - ') -> into an aminoacid sequence (protein)\n",
    "PS: The input should be multiple of 3 (example: 3 nucleotides, 6 nucleotides, etc.)\n",
    "'''\n",
    "\n",
    "from textwrap import wrap\n",
    "def translatedDNA(DNA: str):\n",
    "    \n",
    "    # complement\n",
    "    DNA = DNA.upper()\n",
    "    print(\"DNA input: \", DNA)\n",
    "    nucleotides = {'A','T','C','G'}\n",
    "    if not set(DNA).issubset(nucleotides):\n",
    "        print('Invalid sequence')\n",
    "        return\n",
    "    \n",
    "    length = len(DNA)\n",
    "    if not (length % 3) == 0:\n",
    "        print('Invalid sequence')\n",
    "        return\n",
    "\n",
    "    \n",
    "    compList = {'A': 'T', 'T': 'A', 'C': 'G', 'G': 'C'}\n",
    "    comp = ''.join([compList[base] for base in DNA])\n",
    "    print(\"Complement: \", comp)\n",
    "\n",
    "    mRNA = comp.replace(\"T\", \"U\")\n",
    "    print(\"mRNA: \", mRNA)\n",
    "\n",
    "    # codon table imported from https://gist.github.com/juanfal/09d7fb53bd367742127e17284b9c47bf\n",
    "    codontable = {\n",
    "    'TCA': 'Ser (S) - ',    # Serina\n",
    "    'TCC': 'Ser (S) - ',    # Serina\n",
    "    'TCG': 'Ser (S) - ',    # Serina\n",
    "    'TCT': 'Ser (S) - ',    # Serina\n",
    "    'TTC': 'Phe (F) - ',    # Fenilalanina\n",
    "    'TTT': 'Phe (F) - ',    # Fenilalanina\n",
    "    'TTA': 'Leu (L) - ',    # Leucina\n",
    "    'TTG': 'Leu (L) - ',    # Leucina\n",
    "    'TAC': 'Tyr (Y) - ',    # Tirosina\n",
    "    'TAT': 'Tyr (Y) - ',    # Tirosina\n",
    "    'TAA': '* - ',    # Stop\n",
    "    'TAG': '* - ',    # Stop\n",
    "    'TGC': 'Cys (C) - ',    # Cisteina\n",
    "    'TGT': 'Cys (C) - ',    # Cisteina\n",
    "    'TGA': '* - ',    # Stop\n",
    "    'TGG': 'Trp (W) - ',    # Triptofano\n",
    "    'CTA': 'Leu (L) - ',    # Leucina\n",
    "    'CTC': 'Leu (L) - ',    # Leucina\n",
    "    'CTG': 'Leu (L) - ',    # Leucina\n",
    "    'CTT': 'Leu (L) - ',    # Leucina\n",
    "    'CCA': 'Pro (P) - ',    # Prolina\n",
    "    'CCC': 'Pro (P) - ',    # Prolina\n",
    "    'CCG': 'Pro (P) - ',    # Prolina\n",
    "    'CCT': 'Pro (P) - ',    # Prolina\n",
    "    'CAC': 'His (H) - ',    # Histidina\n",
    "    'CAT': 'His (H) - ',    # Histidina\n",
    "    'CAA': 'Gln (Q) - ',    # Glutamina\n",
    "    'CAG': 'Gln (Q) - ',    # Glutamina\n",
    "    'CGA': 'Arg (R) - ',    # Arginina\n",
    "    'CGC': 'Arg (R) - ',    # Arginina\n",
    "    'CGG': 'Arg (R) - ',    # Arginina\n",
    "    'CGT': 'Arg (R) - ',    # Arginina\n",
    "    'ATA': 'Ile (I) - ',    # Isoleucina\n",
    "    'ATC': 'Ile (I) - ',    # Isoleucina\n",
    "    'ATT': 'Ile (I) - ',    # Isoleucina\n",
    "    'ATG': 'Met (M) - ',    # Methionina\n",
    "    'ACA': 'Thr (T) - ',    # Treonina\n",
    "    'ACC': 'Thr (T) - ',    # Treonina\n",
    "    'ACG': 'Thr (T) - ',    # Treonina\n",
    "    'ACT': 'Thr (T) - ',    # Treonina\n",
    "    'AAC': 'Asn (N) - ',    # Asparagina\n",
    "    'AAT': 'Asn (N) - ',    # Asparagina\n",
    "    'AAA': 'Lys (K) - ',    # Lisina\n",
    "    'AAG': 'Lys (K) - ',    # Lisina\n",
    "    'AGC': 'Ser (S) - ',    # Serina\n",
    "    'AGT': 'Ser (S) - ',    # Serina\n",
    "    'AGA': 'Arg (R) - ',    # Arginina\n",
    "    'AGG': 'Arg (R) - ',    # Arginina\n",
    "    'GTA': 'Val (V) - ',    # Valina\n",
    "    'GTC': 'Val (V) - ',    # Valina\n",
    "    'GTG': 'Val (V) - ',    # Valina\n",
    "    'GTT': 'Val (V) - ',    # Valina\n",
    "    'GCA': 'Ala (A) - ',    # Alanina\n",
    "    'GCC': 'Ala (A) - ',    # Alanina\n",
    "    'GCG': 'Ala (A) - ',    # Alanina\n",
    "    'GCT': 'Ala (A) - ',    # Alanina\n",
    "    'GAC': 'Asp (D) - ',    # Acido Aspartico\n",
    "    'GAT': 'Asp (D) - ',    # Acido Aspartico\n",
    "    'GAA': 'Glu (E) - ',    # Acido Glutamico\n",
    "    'GAG': 'Glu (E) - ',    # Acido Glutamico\n",
    "    'GGA': 'Gly (G) - ',    # Glicina\n",
    "    'GGC': 'Gly (G) - ',    # Glicina\n",
    "    'GGG': 'Gly (G) - ',    # Glicina\n",
    "    'GGT': 'Gly (G) - '   # Glicina\n",
    "}\n",
    "    comp = wrap(comp, 3)\n",
    "\n",
    "    protein = ''.join([codontable[codon] for codon in comp])\n",
    "    protein = protein[:-3]\n",
    "    print('Aminoacid: ', protein)\n",
    "\n",
    "translatedDNA('acgatacgccgacatcggacgttt') # INPUT\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mRNA = UAUUAUUAU\n",
      "UAU = 3\n",
      "═══════•●•═══════\n",
      "mRNA = UAUUAUUAC\n",
      "UAU = 2\n",
      "UAC = 1\n",
      "═══════•●•═══════\n",
      "mRNA = UAUUACUAU\n",
      "UAU = 2\n",
      "UAC = 1\n",
      "═══════•●•═══════\n",
      "mRNA = UAUUACUAC\n",
      "UAU = 1\n",
      "UAC = 2\n",
      "═══════•●•═══════\n",
      "mRNA = UACUAUUAU\n",
      "UAC = 1\n",
      "UAU = 2\n",
      "═══════•●•═══════\n",
      "mRNA = UACUAUUAC\n",
      "UAC = 2\n",
      "UAU = 1\n",
      "═══════•●•═══════\n",
      "mRNA = UACUACUAU\n",
      "UAC = 2\n",
      "UAU = 1\n",
      "═══════•●•═══════\n",
      "mRNA = UACUACUAC\n",
      "UAC = 3\n",
      "═══════•●•═══════\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "Provide the frequency of each RNA codon encoding a given aminoacid, in a DNA \n",
    "sequence\n",
    "PS: \n",
    "• Please provide all possibilities of the codon \n",
    "combination.\n",
    "• Input max. 3 Aminoacids.\n",
    "\n",
    "'''\n",
    "\n",
    "import itertools\n",
    "\n",
    "def mRNA_combinations(protein: str):\n",
    "    codontable = {\n",
    "    'A': ['GCU', 'GCC', 'GCA', 'GCG'],  # Alanine\n",
    "    'C': ['UGU', 'UGC'],                # Cysteine\n",
    "    'D': ['GAU', 'GAC'],                # Aspartic Acid\n",
    "    'E': ['GAA', 'GAG'],                # Glutamic Acid\n",
    "    'F': ['UUU', 'UUC'],                # Phenylalanine\n",
    "    'G': ['GGU', 'GGC', 'GGA', 'GGG'],  # Glycine\n",
    "    'H': ['CAU', 'CAC'],                # Histidine\n",
    "    'I': ['AUU', 'AUC', 'AUA'],         # Isoleucine\n",
    "    'K': ['AAA', 'AAG'],                # Lysine\n",
    "    'L': ['UUA', 'UUG', 'CUA', 'CUC', 'CUG', 'CUU'],  # Leucine\n",
    "    'M': ['AUG'],                       # Methionine\n",
    "    'N': ['AAU', 'AAC'],                # Asparagine\n",
    "    'P': ['CCU', 'CCC', 'CCA', 'CCG'],  # Proline\n",
    "    'Q': ['CAA', 'CAG'],                # Glutamine\n",
    "    'R': ['CGU', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],  # Arginine\n",
    "    'S': ['UCU', 'UCC', 'UCA', 'UCG', 'AGU', 'AGC'],  # Serine\n",
    "    'T': ['ACU', 'ACC', 'ACA', 'ACG'],  # Threonine\n",
    "    'V': ['GUU', 'GUC', 'GUA', 'GUG'],  # Valine\n",
    "    'W': ['UGG'],                       # Tryptophan\n",
    "    'Y': ['UAU', 'UAC'],                # Tyrosine\n",
    "    '*': ['UAA', 'UAG', 'UGA'],         # Stop\n",
    "}\n",
    "\n",
    "\n",
    "    codon_combos = [codontable[aa] for aa in protein]  \n",
    "\n",
    "    # all possible combinations\n",
    "    all_combos = list(itertools.product(*codon_combos))\n",
    "\n",
    "     # count codon frequency for all possible combinations\n",
    "    for combination in all_combos:\n",
    "        mRNA = \"\".join(combination)\n",
    "        codon_count = {}\n",
    "\n",
    "        for codon in combination:\n",
    "            if codon in codon_count:\n",
    "                codon_count[codon] += 1\n",
    "            else:\n",
    "                codon_count[codon] = 1\n",
    "\n",
    "        print(\"mRNA =\", mRNA)\n",
    "        for codon, count in codon_count.items():\n",
    "            print(f\"{codon} = {count}\")\n",
    "        print(\"═══════•●•═══════\")\n",
    "\n",
    "\n",
    "aminoAcid = input(\"Enter amino acids (max 3 letters): \").upper()\n",
    "if len(aminoAcid) > 3:\n",
    "    print(\"Invalid amino acid input\")\n",
    "else:\n",
    "    mRNA_combinations(aminoAcid)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
